Fast parallel-optimized DBSCAN

reductionista opened the pull request:
This is a combination of over 92 commits that have been squashed, and while it works and the performance is great, it's still a "work in progress" in that, we need to add tests for it, clean up some things, and make sure the brute force option still functions properly, along with any other special cases (such as different names for id and point columns, etc.).  Also todo:  clean up and consolidate commit messages, which for now have just been aggregated and pruned a bit.

Opening now so that a PR review can begin for this.

  This is a re-write / 2nd attempt at improving over the original
  brute force implementation. This seems to be working now, and
  should have roughly O(N/S log N) runtimes, where N is the size
  of the input dataset (number of points/rows) and S is the number
  of segments used (note:  the algorithm now decides on its own how
  many segments to use, as sometimes splitting things up further
  will only increase the runtime.  Therefore S is not necessarily
  the total number of segments in the cluster, it may be less
  depending on the structure of the dataset.)

  This borrows many aspects from the previous attempt (using
  an overlapping spatial binary tree to segment the dataset and
  using an R tree index to speed up range queries), but differs in
  its approach in other ways.

  === Rewording of previous summary of the improvements

  The brute force DBSCAN runs on N^2 time. To improve this,
  we split the data into different overlapping regions, running
  DBSCAN on each in parallel, then merging the results together
  on the coordinator. More specifically:

  1. The data is segmented into connected spatial regions, with
     a binary tree optimized specifically for DBSCAN.

  2. Each leaf of the optimized spatial tree runs the dbscan
     algorithm on the points in its spatial region, including
     some points from other regions near its boundaries, using
     an R tree index for efficient range queries.

khannaekta commented:
The new files that were added as part of this PR are failing the license check run as part of the cloudbees build. Please add the license header to all newly added files.

orhankislal reviewed (approved):
I read through the comment and the code and tested a couple of handmade test cases. It seems to work as expected. We should condense the commit message to something more manageable and add these design decisions to the design doc. I noticed some odd split decisions but it might be due to the small dataset sizes (<30 rows, 2d). 
Awesome work, thanks a lot for putting in the time to complete this feature!

khannaekta commented on src/ports/postgres/modules/dbscan/dbscan.py_in:
```diff
@@ -653,25 +651,26 @@ class DBSCAN_optimized(DBSCAN):
             SELECT s.id,
                 ARRAY[i] AS coords,
-                CASE WHEN s.point[i] <= cutoff 
+                CASE WHEN s.point[i] < cutoff
```
Do you think it makes sense to add a dev-check test with a small dataset to be sure we do not break this in future? I will leave it to you, we can skip it if you don't think it will help.

reductionista replied:
Yes, that's a good idea--I'll add one

khannaekta reviewed (commented):
LGTM! except for minor comments. Also pushed 2 commits for fixing the test failure in the Jenkins pipeline and work for GP5.

reductionista commented on src/ports/postgres/modules/dbscan/dbscan.py_in:
```diff
@@ -164,76 +274,1271 @@ def dbscan(schema_madlib, source_table, output_table, id_column, expr_point, eps
             SELECT  '{id_column}'::VARCHAR AS id_column,
+                    '{expr_point}'::VARCHAR AS expr_point,
                     {eps}::DOUBLE PRECISION AS eps,
+                    {min_samples}::BIGINT AS min_points,
```
Hmm... I guess we could do that.

That reminds me, I was planning to ask @orhankislal 's about this.   I'm guessing the main reason why the summary table originally only included `eps` and `metric` is that these are both required inputs for predict to work, while the others are not.  But I'm not sure if there may have been other motivations for keeping the summary table minimal (consistency with other modules?)

I added `min_samples` to this branch when I noticed how difficult its absence made some tasks.  If you have several DBSCAN output tables laying around from a while back, and you want to run it on a new one that's similar to the others (eg. scaled up or down in size with random sampling, noise added, a test/validation split, etc.) but you never wrote down what inputs were used, you have to regenerate all of the outputs again since there is no way of knowing what value of min_samples was used.   While you'll get different outputs for different values of min_samples, you can't know for sure what value was used because it could be that there just weren't any clusters smaller than a certain size.  That could be an important feature of your input data, or it could just be an artifact of the inputs used.

I didn't add `algorithm` or `max_depth` because they are not inputs to DBSCAN in the computational sense of mapping a set of inputs to a set of outputs.  If you know what the input table is, and you know `eps` and `min_samples`, those 3 together determine what the output looks like (aside from some slight randomness in which cluster a point near multiple clusters gets assigned to, due to the order in which points are considered).  Neither `max_depth` nor `algorithm` affect the output, they only affect the runtime.  So my thinking was... there's probably no reason a data scientist would care later by what means the output was generated, as long as they have all the inputs.

I don't have any strong objection to including `algorithm` in the summary, but if we do then I guess we should also include `max_depth` since that also affects the runtime and is sort of like choosing a sub-algorithm among different variants of `optimized`... all of which should give the same result.  There is one use-case I can think of for it, which is performance comparisons of DBSCAN running on similar datasets, such as scale testing.  It would at least be useful for developers, if someone wants to extend DBSCAN or improve it.  And possibly for some end-users, if they wanted to do some runtime-calibration on smaller datasets before attempting something very large.

khannaekta replied:
The only reason I added this was when I ran it with default values, I wasn't sure that it ran with `optimized`/`brute_force`.
I think it's ok for now, let's see if we have any ask for it in the future.

reductionista commented on src/ports/postgres/modules/dbscan/dbscan.py_in:
```diff
@@ -164,76 +274,1271 @@ def dbscan(schema_madlib, source_table, output_table, id_column, expr_point, eps
+                ) params,
+                (
+                    SELECT
+                        {self.schema_madlib}.__dbscan_safe_exp(sum(
+                            {self.schema_madlib}.__dbscan_safe_ln(size)
+                        )) AS V
```
It's equivalent to PROD(size)... the product of all of the values in the size column.  I implemented it this way because I couldn't find any builtin function that does that.  It's possible there is one and I just missed it... if so, that would be cleaner.

reductionista replied:
If there's not, I think I should define this as an actual UDA... I can definitely see why SELECT PROD(size) AS V would be much more readable--glad you brought it up.

reductionista commented on tool/jenkins/jenkins_build.sh:
```diff
@@ -60,9 +60,9 @@ sleep 15
 EOF
-docker exec madlib bash -c 'apt-get update; apt-get install -y python-pip; pip install mock' | tee $workdir/logs/madlib_compile.log
+docker exec madlib bash -c 'apt-get update; apt-get install -y python-pip; pip install mock scipy==1.2.1' | tee $workdir/logs/madlib_compile.log
 
```
I hadn't noticed there was still an import for `scipy` in the DBSCAN module.  That was used in an earlier version, but now we are using `np.linalg.norm` to compute distances instead.  I've removed the dependency from DBSCAN, but we should keep this change in jenkins since kmeans does import `scipy`--not sure how it was working before.

Thanks for this and the gpdb5 port--great to see it working on all platforms now!

reductionista commented:
Addressed most of the comments.  Tasks remaining:  add the tests suggested by @khannaekta , and look into the potential issue @orhankislal pointed out offline.
